File: docs/Archived_Docs/scripts/scraper.ts
/**
 * B2B Lead Intelligence Scraper
 * Monitors news wires and business sources for trigger events
 * Uses Playwright and OpenAI to generate personalized pitches
 */

import { chromium, Browser, Page } from 'playwright'
import { createClient, SupabaseClient } from '@supabase/supabase-js'
import OpenAI from 'openai'

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL!
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY!
const OPENAI_API_KEY = process.env.OPENAI_API_KEY!

if (!SUPABASE_URL || !SUPABASE_SERVICE_KEY) {
  throw new Error('Missing Supabase environment variables')
}

if (!OPENAI_API_KEY) {
  throw new Error('Missing OpenAI API key')
}

const openai = new OpenAI({ apiKey: OPENAI_API_KEY })
const supabase: SupabaseClient = createClient(SUPABASE_URL, SUPABASE_SERVICE_KEY)

interface NewsSource {
  name: string
  url: string
  selectors: {
    articles: string
    title: string
    link: string
    excerpt?: string
  }
}

const NEWS_SOURCES: NewsSource[] = [
  {
    name: 'TechCrunch',
    url: 'https://techcrunch.com',
    selectors: {
      articles: 'article.post-block',
      title: 'h2.post-block__title a',
      link: 'h2.post-block__title a',
      excerpt: '.post-block__content',
    },
  },
  {
    name: 'Crunchbase News',
    url: 'https://news.crunchbase.com',
    selectors: {
      articles: 'article',
      title: 'h2 a',
      link: 'h2 a',
      excerpt: '.excerpt',
    },
  },
]

const TRIGGER_KEYWORDS: Record<string, string[]> = {
  'New Funding': ['funding', 'raised', 'series', 'investment', 'venture capital', 'seed round', 'funded'],
  'New Hires': ['hired', 'appointed', 'joins', 'new executive', 'c-suite', 'new ceo', 'new cto'],
  'Expansion': ['expanding', 'new office', 'entering', 'launching in', 'international', 'global expansion'],
  'Product Launch': ['launched', 'unveils', 'announces', 'new product', 'release', 'rollout'],
  'Partnership': ['partnership', 'partners with', 'collaboration', 'teams up', 'strategic alliance'],
}

class LeadScraper {
  private browser: Browser | null = null

  async init(): Promise<void> {
    this.browser = await chromium.launch({ headless: true })
    console.log('✅ Browser launched')
  }

  async close(): Promise<void> {
    if (this.browser) {
      await this.browser.close()
      console.log('✅ Browser closed')
    }
  }

  async scrapeSource(source: NewsSource): Promise<any[]> {
    if (!this.browser) throw new Error('Browser not initialized')

    const page: Page = await this.browser.newPage()
    const leads: any[] = []

    try {
      console.log(`🔍 Scraping ${source.name}...`)
      await page.goto(source.url, { waitUntil: 'networkidle', timeout: 30000 })
      await page.waitForSelector(source.selectors.articles, { timeout: 10000 })

      const articleElements = await page.$$(source.selectors.articles)

      for (let i = 0; i < Math.min(articleElements.length, 10); i++) {
        const element = articleElements[i]

        try {
          const titleElement = await element.$(source.selectors.title)
          if (!titleElement) continue

          const title = await titleElement.innerText()
          const link = await element.getAttribute(source.selectors.link) || 
                      await titleElement.getAttribute('href') || ''
          const fullLink = link.startsWith('http') ? link : new URL(link, source.url).href

          const excerptElement = source.selectors.excerpt 
            ? await element.$(source.selectors.excerpt)
            : null
          const excerpt = excerptElement ? await excerptElement.innerText() : ''

          // Check if article contains trigger keywords
          const triggerEvent = this.detectTriggerEvent(title + ' ' + excerpt)

          if (triggerEvent) {
            // Extract company name using AI
            const companyName = await this.extractCompanyName(title, excerpt)

            // Check if lead already exists
            const { data: existing } = await supabase
              .from('leads')
              .select('id')
              .eq('company_name', companyName)
              .eq('trigger_event', triggerEvent)
              .order('created_at', { ascending: false })
              .limit(1)
              .single()

            if (existing) {
              console.log(`⏭️  Skipping duplicate: ${companyName}`)
              continue
            }

            // Enrich lead with LinkedIn and Email
            const enrichment = await this.enrichLead(companyName)
            console.log(`🔍 Enriching ${companyName}...`)

            // Generate AI-powered personalized pitch
            const pitch = await this.generatePitch(companyName, triggerEvent, title, excerpt)

            leads.push({
              company_name: companyName,
              trigger_event: triggerEvent,
              contact_email: null, // Legacy field
              prospect_email: enrichment.email,
              prospect_linkedin: enrichment.linkedin,
              ai_personalized_pitch: pitch,
              source_url: fullLink,
            })

            console.log(`✅ Found lead: ${companyName} - ${triggerEvent}`)
          }
        } catch (error) {
          console.error(`Error processing article ${i}:`, error)
          continue
        }
      }
    } catch (error) {
      console.error(`Error scraping ${source.name}:`, error)
    } finally {
      await page.close()
    }

    return leads
  }

  detectTriggerEvent(text: string): string | null {
    const textLower = text.toLowerCase()

    for (const [eventType, keywords] of Object.entries(TRIGGER_KEYWORDS)) {
      if (keywords.some(keyword => textLower.includes(keyword.toLowerCase()))) {
        return eventType
      }
    }

    return null
  }

  async extractCompanyName(title: string, excerpt: string): Promise<string> {
    try {
      const response = await openai.chat.completions.create({
        model: 'gpt-4o',
        messages: [
          {
            role: 'system',
            content: 'Extract the company name from the following news article. Return ONLY the company name, nothing else. If you cannot identify a company, return "Unknown Company".',
          },
          {
            role: 'user',
            content: `Title: ${title}\n\nExcerpt: ${excerpt.substring(0, 300)}`,
          },
        ],
        temperature: 0.3,
        max_tokens: 50,
      })

      return response.choices[0]?.message?.content?.trim() || 'Unknown Company'
    } catch (error) {
      console.error('Error extracting company name:', error)
      // Fallback: use first few words of title
      return title.split(' ').slice(0, 3).join(' ')
    }
  }

  /**
   * Enrichment Engine: Fetch LinkedIn Profile and Verified Email
   * Placeholder implementation for production API integration
   */
  async enrichLead(companyName: string): Promise<{ linkedin?: string; email?: string }> {
    try {
      // PLACEHOLDER: In production, replace with actual enrichment API
      // Examples: Apollo.io, Clearbit, Hunter.io, Lusha, etc.
      
      const enrichmentApiUrl = process.env.ENRICHMENT_API_URL || 'https://api.enrichment-service.com/v1/leads'
      const apiKey = process.env.ENRICHMENT_API_KEY

      if (!apiKey) {
        console.log('⚠️  Enrichment API key not configured, using placeholder data')
        // Return placeholder data for development
        return {
          linkedin: `https://www.linkedin.com/company/${companyName.toLowerCase().replace(/\s+/g, '-')}`,
          email: `contact@${companyName.toLowerCase().replace(/\s+/g, '')}.com`,
        }
      }

      // PLACEHOLDER API CALL - Replace with actual implementation
      const response = await fetch(enrichmentApiUrl, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${apiKey}`,
        },
        body: JSON.stringify({
          company: companyName,
        }),
      })

      if (!response.ok) {
        console.warn(`Enrichment API returned ${response.status}, using placeholder data`)
        return {
          linkedin: `https://www.linkedin.com/company/${companyName.toLowerCase().replace(/\s+/g, '-')}`,
          email: `contact@${companyName.toLowerCase().replace(/\s+/g, '')}.com`,
        }
      }

      const data = await response.json()
      return {
        linkedin: data.linkedin_url || data.linkedin_profile || undefined,
        email: data.verified_email || data.email || undefined,
      }
    } catch (error) {
      console.error('Error enriching lead:', error)
      // Return placeholder data on error
      return {
        linkedin: `https://www.linkedin.com/company/${companyName.toLowerCase().replace(/\s+/g, '-')}`,
        email: `contact@${companyName.toLowerCase().replace(/\s+/g, '')}.com`,
      }
    }
  }

  async generatePitch(
    companyName: string,
    triggerEvent: string,
    title: string,
    excerpt: string
  ): Promise<string> {
    try {
      // Import the centralized AI logic
      const { generatePitch } = await import('../lib/ai-logic')
      const additionalContext = `News Title: ${title}\nContext: ${excerpt.substring(0, 500)}`
      
      return await generatePitch(companyName, triggerEvent, undefined, additionalContext)
    } catch (error) {
      console.error('Error generating pitch:', error)
      // Fallback pitch with correct format
      return `I've already generated a competitive intelligence report for ${companyName} based on your recent ${triggerEvent}. View it here: https://dazrael.com`
    }
  }

  async saveLeads(leads: any[]): Promise<void> {
    if (leads.length === 0) {
      console.log('No new leads to save')
      return
    }

    try {
      const { data, error } = await supabase
        .from('leads')
        .insert(leads)
        .select()

      if (error) throw error

      console.log(`✅ Saved ${leads.length} leads to database`)
    } catch (error) {
      console.error('Error saving leads:', error)
      throw error
    }
  }

  async run(): Promise<void> {
    console.log('🚀 Starting B2B Lead Intelligence Scraper...\n')

    await this.init()

    const allLeads: any[] = []

    try {
      for (const source of NEWS_SOURCES) {
        const leads = await this.scrapeSource(source)
        allLeads.push(...leads)
        await new Promise(resolve => setTimeout(resolve, 2000)) // Rate limiting
      }

      if (allLeads.length > 0) {
        await this.saveLeads(allLeads)
        console.log(`\n✨ Scraping complete! Found ${allLeads.length} new leads.`)
      } else {
        console.log('\n✨ Scraping complete! No new leads found.')
      }
    } catch (error) {
      console.error('Fatal error:', error)
      throw error
    } finally {
      await this.close()
    }
  }
}

// Main execution
async function main() {
  const scraper = new LeadScraper()
  try {
    await scraper.run()
    process.exit(0)
  } catch (error) {
    console.error('Scraper failed:', error)
    process.exit(1)
  }
}

main()

File: docs/Archived_Docs/scripts/scraper.py
"""
B2B Lead Intelligence Scraper
Monitors news wires and business sources for trigger events
"""

import asyncio
import json
import os
from datetime import datetime
from typing import List, Dict, Optional
from playwright.async_api import async_playwright, Browser, Page
from supabase import create_client, Client
import openai

# Configuration
SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY")
OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")

# News sources to monitor
NEWS_SOURCES = [
    {
        "name": "TechCrunch",
        "url": "https://techcrunch.com",
        "selectors": {
            "articles": "article.post-block",
            "title": "h2.post-block__title a",
            "link": "h2.post-block__title a",
            "excerpt": ".post-block__content"
        }
    },
    {
        "name": "Crunchbase News",
        "url": "https://news.crunchbase.com",
        "selectors": {
            "articles": "article",
            "title": "h2 a",
            "link": "h2 a",
            "excerpt": ".excerpt"
        }
    }
]

# Keywords to identify trigger events
TRIGGER_KEYWORDS = {
    "funding": ["funding", "raised", "series", "investment", "venture capital", "seed round"],
    "new_hires": ["hired", "appointed", "joins", "new executive", "c-suite"],
    "expansion": ["expanding", "new office", "entering", "launching in", "international"],
    "product_launch": ["launched", "unveils", "announces", "new product", "release"],
    "partnership": ["partnership", "partners with", "collaboration", "teams up"]
}


class LeadScraper:
    def __init__(self):
        self.supabase: Optional[Client] = None
        if SUPABASE_URL and SUPABASE_KEY:
            self.supabase = create_client(SUPABASE_URL, SUPABASE_KEY)
        self.openai_client = openai.OpenAI(api_key=OPENAI_API_KEY) if OPENAI_API_KEY else None

    async def scrape_source(self, source: Dict, browser: Browser) -> List[Dict]:
        """Scrape a single news source for articles"""
        page = await browser.new_page()
        articles = []

        try:
            print(f"Scraping {source['name']}...")
            await page.goto(source['url'], wait_until="networkidle")
            await page.wait_for_selector(source['selectors']['articles'], timeout=10000)

            article_elements = await page.query_selector_all(source['selectors']['articles'])

            for element in article_elements[:10]:  # Limit to 10 most recent
                try:
                    title_elem = await element.query_selector(source['selectors']['title'])
                    if not title_elem:
                        continue

                    title = await title_elem.inner_text()
                    link = await title_elem.get_attribute('href')
                    
                    if not link.startswith('http'):
                        link = source['url'] + link

                    excerpt_elem = await element.query_selector(source['selectors'].get('excerpt', ''))
                    excerpt = await excerpt_elem.inner_text() if excerpt_elem else ""

                    # Check if article contains trigger keywords
                    event_type = self.detect_trigger_event(title + " " + excerpt)
                    
                    if event_type:
                        # Extract company name using AI
                        company_name = await self.extract_company_name(title, excerpt)
                        
                        articles.append({
                            "title": title.strip(),
                            "link": link,
                            "excerpt": excerpt.strip(),
                            "event_type": event_type,
                            "company_name": company_name,
                            "source": source['name']
                        })
                except Exception as e:
                    print(f"Error processing article: {e}")
                    continue

        except Exception as e:
            print(f"Error scraping {source['name']}: {e}")
        finally:
            await page.close()

        return articles

    def detect_trigger_event(self, text: str) -> Optional[str]:
        """Detect trigger event type from text"""
        text_lower = text.lower()
        
        for event_type, keywords in TRIGGER_KEYWORDS.items():
            if any(keyword in text_lower for keyword in keywords):
                return event_type
        
        return None

    async def extract_company_name(self, title: str, excerpt: str) -> str:
        """Extract company name using OpenAI"""
        if not self.openai_client:
            # Fallback: simple extraction
            words = title.split()
            return " ".join(words[:3]) if len(words) > 3 else title[:50]

        try:
            response = self.openai_client.chat.completions.create(
                model="gpt-4o-mini",
                messages=[
                    {
                        "role": "system",
                        "content": "Extract the company name from the following news article. Return only the company name, nothing else."
                    },
                    {
                        "role": "user",
                        "content": f"Title: {title}\n\nExcerpt: {excerpt[:200]}"
                    }
                ],
                temperature=0.3,
                max_tokens=50
            )
            
            company_name = response.choices[0].message.content.strip()
            return company_name
        except Exception as e:
            print(f"Error extracting company name: {e}")
            return title.split()[0] if title else "Unknown Company"

    async def save_events(self, events: List[Dict]):
        """Save trigger events to Supabase"""
        if not self.supabase:
            print("Supabase not configured. Events not saved.")
            return

        for event in events:
            try:
                # Check if event already exists
                existing = self.supabase.table('trigger_events')\
                    .select('id')\
                    .eq('source_url', event['link'])\
                    .execute()

                if existing.data and len(existing.data) > 0:
                    continue  # Skip duplicates

                # Insert new event
                self.supabase.table('trigger_events').insert({
                    "company_name": event['company_name'],
                    "event_type": event['event_type'],
                    "event_description": f"{event['title']}\n\n{event['excerpt']}",
                    "source_url": event['link'],
                    "detected_at": datetime.utcnow().isoformat(),
                    "company_url": self.extract_company_url(event['link'])
                }).execute()

                print(f"Saved event: {event['company_name']} - {event['event_type']}")
            except Exception as e:
                print(f"Error saving event: {e}")

    def extract_company_url(self, article_url: str) -> Optional[str]:
        """Extract company URL from article (simplified)"""
        # In production, you'd parse the article to find company website
        return None

    async def run(self):
        """Main scraper execution"""
        print("Starting B2B Lead Intelligence Scraper...")
        
        async with async_playwright() as p:
            browser = await p.chromium.launch(headless=True)
            
            all_events = []
            
            for source in NEWS_SOURCES:
                events = await self.scrape_source(source, browser)
                all_events.extend(events)
                await asyncio.sleep(2)  # Rate limiting

            await browser.close()

            print(f"\nFound {len(all_events)} trigger events")
            
            if all_events:
                await self.save_events(all_events)
                print("Events saved to database")
            else:
                print("No trigger events found")

            return all_events


async def main():
    scraper = LeadScraper()
    await scraper.run()


if __name__ == "__main__":
    asyncio.run(main())
